platform/kube/cache.go

```diff
@@ -0,0 +1,138 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kube
+
+import (
+	"istio.io/pilot/model"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+// PodCache is an eventually consistent pod cache
+type PodCache struct {
+	cacheHandler
+
+	// keys maintains stabled pod IP to name key mapping
+	// this allows us to retrieve the latest status by pod IP
+	keys map[string]string
+}
+
+func newPodCache(ch cacheHandler) *PodCache {
+	out := &PodCache{
+		cacheHandler: ch,
+		keys:         make(map[string]string),
+	}
+
+	ch.handler.Append(func(obj interface{}, ev model.Event) error {
+		pod := *obj.(*v1.Pod)
+		ip := pod.Status.PodIP
+		if len(ip) > 0 {
+			switch ev {
+			case model.EventAdd, model.EventUpdate:
+				out.keys[ip] = KeyFunc(pod.Name, pod.Namespace)
+			case model.EventDelete:
+				delete(out.keys, ip)
+			}
+		}
+		return nil
+	})
+	return out
+}
+
+// getPodByIp returns the pod or nil if pod not found or an error occurred
+func (pc *PodCache) getPodByIP(addr string) (*v1.Pod, bool) {
+	key, exists := pc.keys[addr]
+	if !exists {
+		return nil, false
+	}
+	item, exists, err := pc.informer.GetStore().GetByKey(key)
+	if !exists || err != nil {
+		return nil, false
+	}
+	return item.(*v1.Pod), true
+}
+
+// tagsByIP returns pod tags or nil if pod not found or an error occurred
+func (pc *PodCache) tagsByIP(addr string) (model.Tags, bool) {
+	pod, exists := pc.getPodByIP(addr)
+	if !exists {
+		return nil, false
+	}
+	return convertTags(pod.ObjectMeta), true
+}
+
+// NodeCache is an eventually consistent node cache
+type NodeCache struct {
+	cacheHandler
+
+	// keys maintains stable node internal IP to name key mapping
+	// this allows us to retrieve the latest status by node IP
+	keys map[string]string
+}
+
+func newNodeCache(ch cacheHandler) *NodeCache {
+	out := &NodeCache{
+		cacheHandler: ch,
+		keys:         make(map[string]string),
+	}
+
+	ch.handler.Append(func(obj interface{}, ev model.Event) error {
+		node := *obj.(*v1.Node)
+		var ip string
+		for _, a := range node.Status.Addresses {
+			if a.Type == "InternalIP" {
+				ip = a.Address
+				break
+			}
+		}
+		if len(ip) > 0 {
+			switch ev {
+			case model.EventAdd, model.EventUpdate:
+				out.keys[ip] = node.Name
+			case model.EventDelete:
+				delete(out.keys, ip)
+			}
+		}
+		return nil
+	})
+	return out
+}
+
+// getNodeByIP returns the node or nil if pod not found or an error occurred
+func (nc *NodeCache) getNodeByIP(addr string) (*v1.Node, bool) {
```

Comment:
Yeah, that and I wanted to keep it in a similar structure to podcache. Also exists/bool often makes code more readable.